ts/private/ts_project_worker.js

```diff
@@ -0,0 +1,384 @@
+    const builder = ts.createEmitAndSemanticDiagnosticsBuilderProgram(rootNames, options, host, oldProgram, configFileParsingDiagnostics, projectReferences);
+    /** @type {Map} */
+    const emittedFiles = host.emittedFiles = host.emittedFiles || new Map();
+    /** @type {Map} */
+    const emittingMap = host.emittedFilesWeak = host.emittedFilesWeak || new Map();
+
+    const emit = builder.emit;
+    builder.emit = (targetSourceFile, writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers) => {
+        writeFile = writeFile || host.writeFile
+        if (!targetSourceFile) {
+            for (const [path, content] of emittedFiles.entries()) {
+                const sourcePath = emittingMap.get(path);
+                if(!builder.getSourceFile(sourcePath)) {
+                    debuglog(`removing ${path} from emit cache`);
+                    emittedFiles.delete(path);
+                    emittingMap.delete(path);
+                } else {
+                    writeFile(path, content);
+                }
+            }
+        }
+
+        const writeF = (fileName, data, writeByteOrderMark, onError, sourceFiles) => {
+            writeFile(fileName, data, writeByteOrderMark, onError, sourceFiles);
+            emittingMap.set(fileName, sourceFiles[0].fileName)
+            emittedFiles.set(fileName, data)
+        }
+        return emit(targetSourceFile, writeF, cancellationToken, emitOnlyDtsFiles, customTransformers)
+    }
+
+    return builder;
+}
+
+/** @type {ts.FormatDiagnosticsHost} */
+const formatDiagnosticHost = {
+    getCanonicalFileName: (path) => path,
+    getCurrentDirectory: ts.sys.getCurrentDirectory,
+    getNewLine: () => ts.sys.newLine,
+}
+
+function printDiagnostics(diagnostics) {
+    worker.log('')
+    worker.log(ts.formatDiagnostics(diagnostics, formatDiagnosticHost))
+}
+
+
+
+function isAncestorDirectory(ancestor, child) {
+    const ancestorChunks = ancestor.split(path.sep).filter(i => !!i);
+    const childChunks = child.split(path.sep).filter(i => !!i);
+    return ancestorChunks.every((chunk, i) => childChunks[i] === chunk)
+}
+
+/** @type {Map<string, {program: ts.EmitAndSemanticDiagnosticsBuilderProgram, host: ts.WatchCompilerHost, previousInputs?: import("@bazel/worker").Inputs}>} */
+const workers = new Map(); 
+
+function getTsConfigPath(args) {
+    return args[args.indexOf('--project') + 1]
+}
+
+function createProgram(args, initialInputs) {
+    const tsconfig = getTsConfigPath(args);
+    const execRoot = path.resolve(process.cwd(), '..', '..', '..')
+
+    debuglog(`tsconfig: ${tsconfig}`);
+    debuglog(`execroot ${execRoot}`);
+
+    const parsedArgs = ts.parseCommandLine(args)
+
+    const directoryWatchers = new Map();
+    const fileWatchers = new Map();
+
+    const knownInputs = new Set(initialInputs);
+
+    /** @type {ts.System} */
+    const sys = {
+        ...ts.sys,
+        write: worker.log,
```

Comment:
Passing `worker.log` assumes `log` doesn't require any context. Should we put a wrapper (or `.bind`) at the top of the file?

Reply:
It doesn't actually. That'd be unnecessary since `worker` is an object that we get via cjs require. `worker = require(...)` 

Also, `log` is a top-level function that doesn't have a context. https://github.com/bazelbuild/rules_nodejs/blob/621c75f1c2daac39ee7a0896938faf6ca4fcbbb9/packages/worker/src/worker.ts#L17